repose-aggregator/components/services/http-client-service/http-client-service-impl/src/main/scala/org/openrepose/core/services/httpclient/HttpClientDecommissioner.scala

```diff
@@ -0,0 +1,97 @@
+/*
+ * _=_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_=
+ * Repose
+ * _-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-
+ * Copyright (C) 2010 - 2015 Rackspace US, Inc.
+ * _-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * =_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_=_
+ */
+package org.openrepose.core.services.httpclient
+
+import java.util.concurrent.{Executors, ScheduledFuture}
+
+import com.typesafe.scalalogging.slf4j.StrictLogging
+import javax.annotation.PreDestroy
+import javax.inject.Named
+import org.apache.http.impl.client.CloseableHttpClient
+
+import scala.concurrent.duration._
+
+/**
+  * Handles decommissioning of HTTP clients. Decommissioning is a process
+  * in which client usage is tracked and when a client is no longer in use
+  * connections managed by that client are closed and cleaned up. The client
+  * itself is also cleaned up, if necessary.
+  * <p>
+  * This system manages its own background processing since it must only
+  * decommission clients when they are no longer in use. For the sake of
+  * consistency and concurrency, synchronization is used.
+  */
+@Named
+class HttpClientDecommissioner extends HttpClientUserManager with StrictLogging {
+
+  logger.debug("Scheduling HttpClientDecommissioner")
+  private val decomDelay: Duration = 5.seconds
```

Comment:
This also exists as an alternative. 
https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/scheduling/annotation/Scheduled.html

Reply:
I actually ran across that when I was trying to figure out the best way to handle the asynchronicity of decommissioning, but ended up going this route for... some reason...

Oh. I think it was because you have to explicitly configure Spring to enable scheduling via annotations:
https://spring.io/guides/gs/scheduling-tasks/#_enable_scheduling

I'm pretty sure that would just involve creating a configuration class with some annotations for Spring to pick up when it scans for this service, but I choose not go down that rabbit hole since this seemed to work fine.

Reply:
you could also just add it to our existing application context

Reply:
How would you add it to the existing application context? I thought that would just be adding a class annotated with `@Configuration` and `@EnableScheduling` to the package being scanned to create the core context?

Oh. I think I found the answer in your link while I was writing this comment. It looks like you should also be able to do:
```
coreContext.register(ScheduledAnnotationBeanPostProcessor.class)
```

Sure, I'll look into switching over to this.